File: src/main/java/com/mymedia/nagasu/config/GlobalExceptionHandler.java
package com.mymedia.nagasu.config;

import com.mymedia.nagasu.dto.ApiResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.multipart.MaxUploadSizeExceededException;

import java.nio.file.NoSuchFileException;
import java.util.NoSuchElementException;


@RestControllerAdvice
public class GlobalExceptionHandler {
    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    @ExceptionHandler(NoSuchFileException.class)
    public ResponseEntity<ApiResponse<?>> handleNotFound(NoSuchFileException e) {
        logger.warn("File not found: {}", e.getMessage());
        var errorBody = new ApiResponse.Failure<>(e.getMessage() != null ? e.getMessage() : "File not Found");
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorBody);
    }

    @ExceptionHandler(NoSuchElementException.class)
    public ResponseEntity<ApiResponse<?>> handleNoSuchElement(NoSuchElementException e) {
        logger.warn("Element not Found: {}", e.getMessage());
        var errorBody = new ApiResponse.Failure<>(e.getMessage() != null ? e.getMessage() : "Element Not Found");
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorBody);
    }

    @ExceptionHandler(IllegalStateException.class)
    public ResponseEntity<ApiResponse<?>> handleIllegalStateException(IllegalStateException e) {
        logger.warn("StateException: {}", e.getMessage());
        var errorBody = new ApiResponse.Failure<>(e.getMessage() != null ? e.getMessage() : "An internal error occurred");
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorBody);
    }

    @ExceptionHandler(MaxUploadSizeExceededException.class)
    public ResponseEntity<ApiResponse<?>> handleMaxUploadSizeExceededException(MaxUploadSizeExceededException e) {
        logger.warn("MaxUploadSizeExceededException: {}", e.getMessage());
        var errorBody = new ApiResponse.Failure<>(e.getMessage() != null ? e.getMessage() : "Max upload size exceeded");
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorBody);
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ApiResponse<?>> handleIllegalArgumentException(IllegalArgumentException e) {
        logger.warn("IllegalArgumentException: {}", e.getMessage());
        var errorBody = new ApiResponse.Failure<>(e.getMessage() != null ? e.getMessage() : "IllegalArgumentException");
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorBody);
    }

    @ExceptionHandler(RuntimeException.class)
    public ResponseEntity<ApiResponse<?>> handleRuntimeException(RuntimeException e) {
        logger.warn("RuntimeException: {}", e.getMessage());
        var errorBody = new ApiResponse.Failure<>(e.getMessage() != null ? e.getMessage() : "A runtime error occurred");
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorBody);
    }
}

File: src/main/kotlin/com/mymedia/nagasu/config/GlobalExceptionHandler.kt
package com.mymedia.nagasu.config

import com.mymedia.nagasu.dto.ApiResponse
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.multipart.MaxUploadSizeExceededException
import java.nio.file.NoSuchFileException

@RestControllerAdvice
class GlobalExceptionHandler {
    private val logger = LoggerFactory.getLogger(javaClass)

    @ExceptionHandler(NoSuchFileException::class)
    fun handleNotFound(e: NoSuchFileException): ResponseEntity<ApiResponse<Nothing>> {
        logger.warn("File not found: ${e.message}")
        val errorBody = ApiResponse.Failure(e.message ?: "File not found")
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorBody)
    }

    @ExceptionHandler(NoSuchElementException::class)
    fun handleNoSuchElement(e: NoSuchElementException): ResponseEntity<ApiResponse<Nothing>> {
        logger.warn("Element not found: ${e.message}")
        val errorBody = ApiResponse.Failure(e.message ?: "Element not found")
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorBody)
    }

    @ExceptionHandler(IllegalStateException::class)
    fun handleIllegalStateException(e: IllegalStateException): ResponseEntity<ApiResponse<Nothing>> {
        logger.warn("StateException: ${e.message}")
        val errorBody = ApiResponse.Failure(e.message ?: "An internal error occurred")
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorBody)
    }

    @ExceptionHandler(MaxUploadSizeExceededException::class)
    fun handleMaxUploadSizeExceededException(e: MaxUploadSizeExceededException): ResponseEntity<ApiResponse<Nothing>> {
        logger.warn("MaxUploadSizeExceededException: ${e.message}")
        val errorBody = ApiResponse.Failure(e.message ?: "Max upload size exceeded")
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorBody)
    }

    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgumentException(e: IllegalArgumentException): ResponseEntity<ApiResponse<Nothing>> {
        logger.warn("IllegalArgumentException: ${e.message}")
        val errorBody = ApiResponse.Failure(e.message ?: "IllegalArgumentException")
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorBody)
    }
}